tests/ui/test_ui_tools.py

```diff
@@ -1804,7 +1811,9 @@ def test_main_view_content_header_without_header(self, mocker, message,
         all_to_vary = dict(to_vary_in_last_message, **stars['last'])
         last_msg = dict(message, **all_to_vary)
         msg_box = MessageBox(this_msg, self.model, last_msg)
-        expected_header[1] = msg_box._time_for_message(message)
+        self.model.formatted_local_time.side_effect = self.formatted_local_time
+        expected_header[1] = self.model.formatted_local_time(
```

Comment:
Initially a few tests were failing when I replaced the `msg_box._time_for_message(message)` with the mocked function call to `self.model.formatted_local_time`. The `self.model` in the class `TestMessageBox` is a MagicMock object and it only returns a mock object when `self.model.formatted_local_time` is called. As a result, we do not get the required return value of type string.
To rectify this, I have set the `.side_effect` of `self.model.formatted_local_time` to another function `formatted_local_time` within the same class `TestMessageBox`. On doing this, the `formatted_local_time` method within the class is accessed whenever a call to the mocked function `self.model.formatted_local_time` is made. The passed arguments are processed to return the required string in this case.

I have repeated the same changes in `test_popups.py` where a similar issue was occuring.

Reply:
We shouldn't be including these duplicates in the tests in any case; that's not a good solution - though I can see your problem.

Previously the test simply used `ctime` internally, however with the big model mock it cannot access the original method. There are two alternatives, in addition to the 'reimplement part of model' approach you have tried already:
* don't mock all of model, only the parts that are required; that could involve quite a refactor before doing this
* set a `return_value` instead of a `side_effect`, custom to each test parametrization; rather than working with timestamps, we then just work with the resulting text we expect to see, and would get from the new function

I'd recommend trying the last option first.